### PR TITLE
Worknet plugins/extensions

### PR DESCRIPTION
**Pull request by ngdentdev**

### CODE REVIEW

**Review by JohndeVadoss — APPROVED** [review]

File: src/worknet-ext/logger/WorkNetLogger.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+using Neo.Ledger;
+using Neo.Network.P2P.Payloads;
+using Neo.Persistence;
+using Neo.Plugins;
+using Neo.SmartContract;
+using Neo.SmartContract.Native;
+
+namespace WorkNetExt;
+
+public class WorkNetLogger : Plugin
```
Comment (devhawk) [inline src/worknet-ext/logger/WorkNetLogger.cs:13]: I get this is a sample plugin, but is there some way we could do something that doesn't completely duplicate what WorkNet's built in log plugin does? Write to a file maybe?

File: src/worknet-ext/logger/WorkNetLogger.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+{
+    NeoSystem? neoSystem;
```
Comment (devhawk) [inline src/worknet-ext/logger/WorkNetLogger.cs:15]: Plugins can override the `Configure` method and call the `GetConfiguration` protected method to retrieve configuration information (in lieu of configuring in code). Here's an [example from the TokenTracker plugin](https://github.com/neo-project/neo-modules/blob/master/src/TokensTracker/TokensTracker.cs#L49): 

```cs
protected override void Configure()
{
    IConfigurationSection config = GetConfiguration();
    _dbPath = config.GetValue("DBPath", "TokensBalanceData");
    _shouldTrackHistory = config.GetValue("TrackHistory", true);
    _maxResults = config.GetValue("MaxResults", 1000u);
    _network = config.GetValue("Network", 860833102u);
    _enabledTrackers = config.GetSection("EnabledTrackers").GetChildren().Select(p => p.Value).ToArray();
}
```

Our sample plugin should demonstrate using Configure/GetConfiguration.

Note, `GetConfiguration` reads from the virtual `ConfigFile` property which (eventually) reads from the static `PluginsDirectory` property. The `PluginsDirectory` property - like the main plugin loaded - is relative to the assembly location. I suspect that worknet compatible plugins will need to overrride `ConfigFile` to read from a location that is relative to the two locations we are loading plugins from.

Comment (devhawk) [reply src/worknet-ext/logger/WorkNetLogger.cs:15]: One other thing, [Neo 3.6](https://github.com/neo-project/neo/issues/2844) is likely to ship soon. It might be worth updating the logic of the platform `Plugin` class. In particular, making the platform `Plugin.LoadPlugins` method public + take an optional directory parameter. Then maybe we can also modify the `ConfigFile/RootPath/PluginsDirectory` properties to track the folder the plugin was originally loaded from.

File: src/worknet/Node/PluginHandler.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            if (type.IsAbstract) continue;
+
+            ConstructorInfo? constructor = type.GetConstructor(Type.EmptyTypes);
+            try
+            {
+                constructor?.Invoke(null);
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"Failed to load plugin: {type.FullName}", ex);
```
Comment (devhawk) [inline src/worknet/Node/PluginHandler.cs:42]: THe platform `LoadPlugin` logs an error on exception instead of throwing. Why did you change this?

File: src/worknet/Node/PluginHandler.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
+using System.Reflection;
+using Neo.Plugins;
+using McMaster.Extensions.CommandLineUtils;
+
+class PluginHandler
+{
+    public static void LoadPlugins(string directory, IConsole console)
```
Comment (devhawk) [inline src/worknet/Node/PluginHandler.cs:7]: Since IConsole is a type specific to the command line library WorkNet uses, we should pass this as a TextWriter instead. Maybe we make it optional instead? Both of these will make it easier to update platform logic and eliminate the need for this class entirely.

**Review by devhawk — CHANGES_REQUESTED** [review]
I made a few comments with suggested changes

File: src/worknet/Node/PluginHandler.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
+using System.Reflection;
+using Neo.Plugins;
+using McMaster.Extensions.CommandLineUtils;
+
+class PluginHandler
```
Comment (devhawk) [inline src/worknet/Node/PluginHandler.cs:5]: static class?

File: src/worknet/Node/PluginHandler.cs (hunk 3 above)
Comment (ngdentdev) [reply src/worknet/Node/PluginHandler.cs]: I did that during debugging, replaced the code to log now

File: src/worknet/Node/PluginHandler.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            }
+            catch (Exception ex)
+            {
+                Log(writer, $"Failed to load plugin: {type.FullName}");
+                Log(writer, ex.Message);
+            }
+        }
+    }
+
+    private static void Log(TextWriter? writer, string message)
```
Comment (devhawk) [inline src/worknet/Node/PluginHandler.cs:47]: Instead of adding a Log method, you can use the [null conditional operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-). Everywhere you currently call Log, use this instead:

```cs
writer?.WriteLine("whatever the message is")
```

File: src/worknet-ext/logger/WorkNetLogger.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    public override void Dispose()
+    {
+        Neo.Utility.Logging -= OnNeoUtilityLog;
+        ApplicationEngine.Log -= OnAppEngineLog!;
+        Blockchain.Committing -= OnCommitting;
+        GC.SuppressFinalize(this);
+    }
+
+    // Overwrite Config file method to find the config.json from the same directory as the plugin dll directory
```
Comment (devhawk) [inline src/worknet-ext/logger/WorkNetLogger.cs:34]: Please expand this comment so that plug in developers undertand that overriding ConfigFile is an absolute requirement for building plugins that are worknet compatible

**Review by devhawk — APPROVED** [review]
ready to merge once null conditional operator use + expanded ConfigFile overload comment are addressed

File: src/worknet/Node/PluginHandler.cs (hunk 6 above)
Comment (ngdentdev) [reply src/worknet/Node/PluginHandler.cs:47]: oh good, didn't know c# had this as well.

**Comment by devhawk** [comment]
:shipit:
